lib/dataStructures/queue.js

```diff
@@ -28,7 +28,12 @@
         };
 
         self.dequeue = function() {
-            return !this.isEmpty() ? this.elements.shift() : null;
+            if (!this.isEmpty()) {
+                var first = this.elements[0];
+                this.elements = this.elements.slice(1);
```

Comment:
I believe `.slice()` falls into similar time complexity as `.shift()`, which is why an offset needs to be maintained so that slice is only called when the underlying array "reaches the halfway mark" (which is far less frequent than every array dequeue operation)